### DISCUSSION TITLE
Why preview mode runs rsbuild flow anyway?

### DISCUSSION THREAD

**Discussion by JounQin** [web-infra-dev/rspress#1458] [Q&A]
https://github.com/web-infra-dev/rspress/blob/e2ebf7a26cb19f5f8230c39b09aa62f7336b5752/packages/cli/src/index.ts#L140

There is no `root` argument to override `config.root` in `preview` command like `build`, so I'd expect that `preview` should only serve `doc_build` dir or`config.outDir` directly.

https://github.com/web-infra-dev/rspress/blob/e2ebf7a26cb19f5f8230c39b09aa62f7336b5752/packages/core/src/node/serve.ts#L36-L41

I don't follow why the whole rsbuild flow should be invoked, that seems just wasting?

**Comment by Timeless0911** [comment]
Using existing preview features of Rsbuild can make it easy to maintain, you can serve it yourself on your demand.

**Reply by JounQin** [reply]
As I mentioned, `preview` command does not provide `[root]` argument, so it could using different `root` when running unnecessary plugins.

<img width="947" alt="image" src="https://github.com/user-attachments/assets/434abaaa-e9d9-4d7a-8b35-8683595d907d">

Of course, I can use a fake `root: cwd` but it seems strange.

I mean we should disable plugins on previewing maybe.

**Reply by Timeless0911** [reply]
Does adding a `--root` cli param to preview command help?

**Reply by JounQin** [reply]
I don't know the implementation details yet, but as I understand, `root`/`plugins` in `preview` mode does nothing actually, because it's just a static server, so adding `[root]` argument is redundant. But I believe disabling rspress plugins would help speeding up preview mode a bit and reduce such confusions.

**Reply by Timeless0911** [reply]
Is there a minimal reproduce repository?

**Reply by chenjiahan** [reply]
I do not agree with removing all plugins during the preview. Running plugins to modify the config during preview is allowed, and it does not have a significant performance impact. I checked that the Vite plugin also runs during vite preview.

**Reply by JounQin** [reply]
> Running plugins to modify the config during preview is allowed

I don't quite understand the use case, I was thinking `preview` only serves static `doc_build` dir, anything else? If so, no `[root]` argument provided may cause inconsistent behavior with `build` command.

---

I noticed this issue because it says `root` is optional in `config` which defaults as `docs` at https://rspress.dev/api/config/config-basic#root, but actually it's required in some core plugins like `@rspress/plugin-auto-nav-sidebar` as I mentioned at https://github.com/web-infra-dev/rspress/discussions/1458#discussioncomment-10901538

The reproduction is just simple: https://github.com/JounQin/rspress-preview-issue/blob/main/cli.ts

**Reply by Timeless0911** [reply]
This may be a bug related to `config.root`, I will fix it later and add a `[root]` argument to preview command.

**Reply by Timeless0911** [reply]
You can use `rspress@0.0.0-next-20241010114926` to verify the behaviour now, the PR is https://github.com/web-infra-dev/rspress/pull/1461
